### PR TITLE
adds melee aim correction

### PR DESCRIPTION
**Pull request by zeskorion**
## About The Pull Request
on harm intent, if you hit a turf or item on harm intent, and a living creature is on it, you will hit that creature. 
this is nearly exactly the same as #2444 , but it doesn't need the user to be walking

## Why It's Good For The Game
byond pixel searching and weird attack intricacies are bogus. This very slightly evens the playing field as far as ping is concerned, and makes cosmetics and being bigger or smaller not as big a deal

inb4 "skill issue"

## Changelog
:cl:
add: On harm intent, if you miss an attack by a few pixels, you'll now hit instead
/:cl:

### CODE REVIEW

**Comment by zeskorion** [comment]
this feels like too little code for what it does

File: code/_onclick/item_attack.dm
Comment (MNarath1) [inline code/_onclick/item_attack.dm]:
```suggestion
			if(L.stat || user == L)
				continue
```

File: code/game/objects/objs.dm
Comment (MNarath1) [inline code/game/objects/objs.dm]:
```suggestion
			if(L.stat || user == L)
				continue
```

File: code/game/turfs/turf.dm
Comment (MNarath1) [inline code/game/turfs/turf.dm]:
```suggestion
			if(L.stat || user == L)
				continue
```

File: code/game/turfs/turf.dm
Comment (MNarath1) [inline code/game/turfs/turf.dm]:
```suggestion
			if(L.stat || user == L)
				continue
```

**Review by MNarath1 — CHANGES_REQUESTED** [review]
I would say combine these checks into one if no reason to call if twice here.

**Review by MNarath1 — COMMENTED** [review]
Also what happens if there is like multiple people resting / in crit on a turf ? Maybe also include people that rest / are in crit in this checks  they are really not hard to hit and the one guy standing above them should be the priority for anyone.

**Comment by VictorVow** [comment]
Helpful change especially for newbies + antags and gets us away from snowflake game behaviours.

**Comment by Archanial** [comment]
This makes combat feel very dull and uninteresting.
🔑

**Comment by FriendlyContractor** [comment]
Trying to click tiny sprites rapidly moving around isnt excatly the best part of the game. I think its a good change

**Comment by zeskorion** [comment]
> code

Resting != in crit. Mobility flags arent accounted for. People who are in crit arent included, stunned people are

**Comment by PowerfulBacon** [comment]
I don't know how I feel about this, I neither want to be for or against this change. I remain neutral.

**Comment by PowerfulBacon** [comment]
The turf one is fine, the item one makes little sense.
If you are on harm intent and try to pick something up, you just end up attacking the person standing on the item.
What about the nuances such as trying to crowbar up a tile, if someone is on it and you are on harm intent then you attack them etc.

**Review by PowerfulBacon — COMMENTED** [review]
See above comment, also does attackby check range or is range checked before hand. If it is the latter, then multi-turf objects will allow you to attack at ranges

**Comment by Kapu1178** [comment]
Goon has a very similar system and it makes the game feel so much better.

**Comment by Effemu** [comment]
Skill issue the pr

**Comment by zeskorion** [comment]
> The turf one is fine, the item one makes little sense. If you are on harm intent and try to pick something up, you just end up attacking the person standing on the item. What about the nuances such as trying to crowbar up a tile, if someone is on it and you are on harm intent then you attack them etc.

People should generally watch their intents anyways. The change is well worth the edge cases



> See above comment, also does attackby check range or is range checked before hand. If it is the latter, then multi-turf objects will allow you to attack at ranges

Will check when i get home

**Comment by zeskorion** [comment]
> See above comment, also does attackby check range or is range checked before hand. If it is the latter, then multi-turf objects will allow you to attack at ranges

attackby checks aiming.

**Comment by MCterra10** [comment]
literally skill issue

**Comment by MCterra10** [comment]
> Goon has a very similar system and it makes the game feel so much better.

goon's system is nothing like this. goon's system is a swing-based attack feature rather than an aim assist, where you perform an attack regardless of if someone is on the tile or not, and can click any tile in the general direction you want to attack, and will swing at the adjacent tile in that direction.

**Comment by ghost** [comment]
Until somebody reworks melee combat entirely, I would argue this is an improvement over the current system of "spam clicks and joust back and forth and around eachother until one of you lands a hit" melee fights revolve around now. 
A combat system where 90%+ of your attacks don't actually hit anything is a terrible system.

**Comment by ghost** [comment]
image

**Comment by mcmeiler** [comment]
literally skill issue

**Comment by MCterra10** [comment]
> This makes combat feel very dull and uninteresting. 🔑

🔑

**Comment by ghost** [comment]
I think it'd be preferred if weapons/attacks had a pixel radius rather than a blanket hit on anything in the turf. Hell, you could give long pokey sticks small radius and big un-pokey hammers big radius. But that'd be more than 29 lines.

**Comment by Tsar-Salat** [comment]
`A combat system where 90%+ of your attacks don't actually hit anything is a terrible system`
Skill issue

**Comment by KubeRoot** [comment]
I think this should be worth trying for a while... Maybe a week, maybe a month?

Anyways, once this is merged we're getting hitscan guns, yeah?

**Comment by zeskorion** [comment]
> I think this should be worth trying for a while... Maybe a week, maybe a month?
> 
> Anyways, once this is merged we're getting hitscan guns, yeah?

Thats for you to decide

**Comment by SergeiKoralev** [comment]
This would be nice. It removes the abuse of figuring out ways to make yourself stupidly small. Hitting a dwarf with a crowbar is not difficult, it's like playing golf.

I do have to agree with Bacon that not being able to pick things up on harm intent is a major downside, but at the same time it could also be considered a balance move since hitting is now easier.

**Comment by Wilchenx** [comment]
> This would be nice. It removes the abuse of figuring out ways to make yourself stupidly small. Hitting a dwarf with a crowbar is not difficult, it's like playing golf.
> 
> I do have to agree with Bacon that not being able to pick things up on harm intent is a major downside, but at the same time it could also be considered a balance move since hitting is now easier.

just alt+click the tile man

**Comment by MNarath1** [comment]
> > This would be nice. It removes the abuse of figuring out ways to make yourself stupidly small. Hitting a dwarf with a crowbar is not difficult, it's like playing golf.
> > I do have to agree with Bacon that not being able to pick things up on harm intent is a major downside, but at the same time it could also be considered a balance move since hitting is now easier.
> 
> just alt+click the tile man

i am pretty sure trying to pick up the item like that would also call `attackby` so the only way to pick up stuff when you are on harm intend would be with the right click menu and that is indeed kinda inconvenient.

**Comment by zeskorion** [comment]
> This would be nice. It removes the abuse of figuring out ways to make yourself stupidly small. Hitting a dwarf with a crowbar is not difficult, it's like playing golf.
> 
> I do have to agree with Bacon that not being able to pick things up on harm intent is a major downside, but at the same time it could also be considered a balance move since hitting is now easier.

You actually still pick up the item you click at the moment

**Comment by The-Moon-Itself** [comment]
> This would be nice. It removes the abuse of figuring out ways to make yourself stupidly small. Hitting a dwarf with a crowbar is not difficult, it's like playing golf.
> 
> I do have to agree with Bacon that not being able to pick things up on harm intent is a major downside, but at the same time it could also be considered a balance move since hitting is now easier.

This is the main reason why I support this. If combat is not dependent on the physical size of the sprite, then it allows for implementation of making people smaller without the potential of powergaming.

**Comment by PowerfulBacon** [comment]
I think this isn't that good, goons is way better where if you click north the attack goes north (that also solves the problem of picking up items attacking people which is going to be SUPER annoying)

**Comment by PowerfulBacon** [comment]
also skill issue

**Comment by PowerfulBacon** [comment]
> > This would be nice. It removes the abuse of figuring out ways to make yourself stupidly small. Hitting a dwarf with a crowbar is not difficult, it's like playing golf.
> > I do have to agree with Bacon that not being able to pick things up on harm intent is a major downside, but at the same time it could also be considered a balance move since hitting is now easier.
> 
> just alt+click the tile man

Alt clicking the tile and using the stat panel to pick up items would still attack players on harm intent.

**Comment by zeskorion** [comment]
> I think this isn't that good, goons is way better where if you click north the attack goes north (that also solves the problem of picking up items attacking people which is going to be SUPER annoying)

I could do this. Making it less simple opens the way for a deal of bugginess, but i could.

**Comment by KubeRoot** [comment]
If kept in current form, I think it shouldn't use `get_turf` and instead check if `loc` is a turf; I think right now if you're on the same turf as somebody else and click an item in your inventory, you'll attack them.

**Comment by Victor239** [comment]
I don't really see the issue with Harm intent preventing picking up items - it's after all in the name of the intent of itself, you're intending to _harm_ with your clicks.

Also if we ever remove intents and go to Combat Mode like other codebases then this would still fit, where one intent allows picking up items whereas the other one attacks the tile.

File: code/game/objects/objs.dm
Comment (utilbird) [inline code/game/objects/objs.dm:385]:
```suggestion
	return ..()
```

Comment (zeskorion) [reply code/game/objects/objs.dm:385]: this is how the attackhand code is elsewhere and im scared its gonna break something esoteric if i change it
(also not a maintainer, i care not)

**Comment by Raven-Industries** [comment]
How's this work with weapons meant to be used in non-harm intents (stun batons, mainly)? It seems like as it is now harmbatoning would have an advantage over standard baton use.

**Comment by zeskorion** [comment]
I may make this pr again at some point, but right now i dont feel like working on it, and it needs to be more fully implemented than i have here

**Pull request closed without merging**
